Main/HW27/HW27/Vector2d.cpp

```diff
@@ -0,0 +1,97 @@
+#include "Vector2d.h"
+
+int Vector2d::count = 0;
+
+Vector2d::Vector2d(float x, float y) : x(x), y(y) {
+    count++;
+}
+
+Vector2d::Vector2d(float x0, float y0, float x1, float y1)
+    : x(x1 - x0), y(y1 - y0) {
+    count++;
+}
+
+Vector2d::Vector2d(const Vector2d& other) : x(other.x), y(other.y) {
+    count++;
+}
+
+Vector2d::~Vector2d() {
+    count--;
+}
+
+Vector2d& Vector2d::operator=(const Vector2d& other) {
+    if (this != &other) {
+        x = other.x;
+        y = other.y;
+    }
+    return *this;
+}
+
+void Vector2d::operator*=(float scalar) {
+    x *= scalar;
+    y *= scalar;
+}
+
+float Vector2d::operator()() const {
+    return std::sqrt(x * x + y * y);
+}
+
+float& Vector2d::operator[](std::size_t idx) {
+    if (idx == 0) return x;
+    else if (idx == 1) return y;
+    else throw std::out_of_range("Index out of range");
+}
+
+float Vector2d::dotProduct(const Vector2d& other) const {
+    return x * other.x + y * other.y;
+}
+
+Vector2d Vector2d::negate() const {
+    return Vector2d(-x, -y);
+}
+
+VectorRelativeState Vector2d::getRelativeState(const Vector2d& other) const {
+    float dot = dotProduct(other);
+    float cross = x * other.y - y * other.x;
```

Comment:
what does cross mean for 2d vector? 😄 
all you need for all those checks are dotProduct direct comparison with operator== to look for Identical

Reply:
See Lecture slide for reference, 5 out of 6 cases are covered there with just dotProduct value